mrmustard/math/lattice/strategies/beamsplitter.py

```diff
@@ -154,3 +154,51 @@ def beamsplitter_vjp(
     dLdphi = 2 * np.real(1j * st * em * dLdA[0, 3] + 1j * st * e * dLdA[1, 2])
 
     return dLdtheta, dLdphi
+
+
+def beamsplitter_schwinger(shape, theta, phi, max_N=None):
+    r"""Returns the Fock representation of the beamsplitter up to
+    the given cutoff for each of the two modes.
+
+    This implementation is in pure python (so it's slower than the numba version),
+    but it's numerically stable up to arbitrary cutoffs.
+
+    In this implementation we split the two-mode Fock basis into finite subsets spanned
+    by |m,n> with m+n=const, i.e. {|0,0>}, {|1,0>,|0,1>}, {|2,0>, |1,1>, |0,2>}, etc...
+    A beamsplitter acts unitariliy in each of these subspaces without mixing them with each other,
+    i.e. in this basis the beamsplitter would be a block-diagonal matrix.
+    This means we can construct the BS matrix by first calculating the BS unitaries
+    in each of these subspaces.
+    This can be done using the matrix exponential, which is numerically stable.
+    We couldn't do this in the original basis because it was infinite-dimensional
+    and we had to truncate it at some point.
+
+    Arguments:
+        shape (int, int, int, int): The shape of the output tensor. Only shapes of the form (i,k,i,k) are supported.
+        theta (float): The angle of the beamsplitter.
+        phi (float): The phase of the beamsplitter.
+        max_N (int): The maximum total photon number to include in the calculation.
+
+    Returns:
+        np.ndarray: The beamsplitter in the Fock basis.
+    """
+    c1, c2, c3, c4 = shape
+    if c1 != c2 or c3 != c4:
+        raise ValueError("The Schwinger method only supports shapes of the form (i,k,i,k).")
```

Comment:
if here would be (i,i,k,k)? Because you are checking c1 != c2 and c3 != c4.

Reply:
oh thank you!